src/puppetlabs/trapperkeeper/core.clj

```diff
@@ -154,33 +157,74 @@
     ;; it can be referenced in a remote nREPL session, etc.
     (swap! internal/tk-apps conj app)
     (internal/register-sighup-handler)
-    (run-app app)
-    (swap! internal/tk-apps (partial remove #{app}))))
+    (let [{:keys [::exit] :as result} (run-app app)]
+      ;; Q: If it's appropriate (or even just acceptable) for this
+      ;; removal to be handled via catch/finally, then it'd be a bit
+      ;; simpler to just throw the exit from run-app.
+      (swap! internal/tk-apps (partial remove #{app}))
+      (when exit
+        (throw (ex-info (i18n/trs "Process exit requested")
+                        (assoc (::exit result) :kind ::exit)))))))
+
+(defn- parse-args [args]
+  "Returns valid CLIData or throws an ::exit."
+  (letfn [(quit [status msg stream ex-msg]
+            (throw (ex-info ex-msg
+                            ;; :status and :messages match exit-request-schema
+                            {:kind ::exit
+                             :status status
+                             :messages [[(str msg "\n") stream]]})))]
+    (try
+      (internal/parse-cli-args! (or args ()))
+      (catch ExceptionInfo ex
+        (let [{:keys [kind msg] :as data} (ex-data ex)]
+          (case (some-> kind without-ns)
+            :cli-error (quit 1 msg *err* (i18n/trs "Invalid program arguments"))
+            :cli-help (quit 0 msg *out* (i18n/trs "Command line --help requested"))
+            (throw ex)))))))
+
+(defn- handle-exit
+  "Prints any messages provided to the indicated streams, and returns
+  the appropriate process exit status."
+  [{:keys [status messages]}]
+  ;; Check values carefully here since throwing ::exit is a public interface
+  (letfn [(show [stream msg]
+            (binding [*out* stream] (print msg) (flush)))]
+    (doseq [[string stream :as m] messages
+            :when (try
```

Comment:
Why is `:when` a keyword here?

Reply:
`doseq` allows you to put in bindings to filter stuff, see the docs for [for](https://clojuredocs.org/clojure.core/for).

Reply:
ahh I see, thank you!

Reply:
Oh, it's one of the "modifiers" supported by `for`: https://clojuredocs.org/clojure.core/for